### PR TITLE
Indirect Interfaces need to work with Project Recovery after a crash

### LINKED ISSUE #23144: Indirect Inelastic and project recovery
The indirect inelastic GUIs can be unstable. To help minimize the problems this causes we want the GUI's to work with project recovery. 

To do this we will have to make sure that algorithms are not set to child, but will have to turn off the store to ADS. See https://github.com/mantidproject/mantid/blob/master/scripts/Muon/GUI/FrequencyDomainAnalysis/MaxEnt/maxent_model.py#L64 

In the` Indirect` folder of `scientific_interfaces` every instance of the word `child` is below (will not need to edit all of them):

```
AbsorptionCorrections.cpp:21:  convertAlg->setChild(true);

AbsorptionCorrections.cpp:37:  groupAlg->setChild(true);

ContainerSubtraction.cpp:470:  shift->setChild(true);

ContainerSubtraction.cpp:484:  scale->setChild(true);

ContainerSubtraction.cpp:498:  minus->setChild(true);

ContainerSubtraction.cpp:512:  rebin->setChild(true);

ContainerSubtraction.cpp:525:  convert->setChild(true);

ConvFitModel.cpp:111:  loadParamFile->setChild(true);

ConvFitModel.cpp:171:  cloneAlg->setChild(true);

ConvFitModel.cpp:186:  appendAlg->setChild(true);

ConvFit.ui:22:     <property name="childrenCollapsible">

CorrectionsTab.cpp:26: * Calls overridden version of loadSettings() in child class.

CorrectionsTab.h:78:  /// Overidden by child class.

CorrectionsTab.h:80:  /// Overidden by child class.

CorrectionsTab.h:82:  /// Overidden by child class.

CorrectionsTab.h:85:  /// Overidden by child class.

IndirectBayes.h:18:interface window and handles the interaction between the child tabs on the

IndirectCorrections.h:59:  /// Load the settings of the interface (and child tabs).

IndirectDataAnalysis.h:58:  /// Load the settings of the interface (and child tabs).

IndirectDataAnalysisTab.cpp:39: * Calls overridden version of loadSettings() in child class.

IndirectDataAnalysisTab.h:136:  /// Overidden by child class.

IndirectDataAnalysisTab.h:138:  /// Overidden by child class.

IndirectDataAnalysisTab.h:140:  /// Overidden by child class.

IndirectDataAnalysisTab.h:143:  /// Overidden by child class.

IndirectDataReduction.cpp:211:    loadAlg->setChild(true);

IndirectDataReduction.cpp:226:      loadParamAlg->setChild(true);

IndirectDataReductionTab.cpp:153:  createWsAlg->setChild(true);

IndirectDataReductionTab.cpp:166:  convertHistAlg->setChild(true);

IndirectDataReductionTab.cpp:175:  loadInstAlg->setChild(true);

IndirectDataReductionTab.cpp:191:  loadParamAlg->setChild(true);

IndirectDataReductionTab.cpp:211:  convUnitsAlg->setChild(true);

IndirectDiffractionReduction.cpp:599:  loadAlg->setChild(true);

IndirectDiffractionReduction.cpp:612:    loadParamAlg->setChild(true);

IndirectFitAnalysisTab.h:184:  /// Overidden by child class.

IndirectFitPlotModel.cpp:310:  createWsAlg->setChild(true);

IndirectFitPlotModel.cpp:326:  extractSpectraAlg->setChild(true);

IndirectFitPlotModel.cpp:343:  appendSpectraAlg->setChild(true);

IndirectFitPlotModel.cpp:358:  cropAlg->setChild(true);

IndirectFitPlotModel.cpp:374:  deleteWorkspaceAlg->setChild(true);

IndirectSimulation.h:19:  handles the interaction between the child tabs on the window.

IndirectTab.h:225:  /// Overidden by child class.

IndirectTab.h:227:  /// Overidden by child class.

IndirectTab.h:229:  /// Overidden by child class.

IndirectTools.h:19:            handles the interaction between the child tabs on the window.

Iqt.cpp:36:    toIqt->setChild(true); // record this as internal

Iqt.cpp:50:    deleteAlg->setChild(true);

IqtFit.ui:22:     <property name="childrenCollapsible">

JumpFitModel.cpp:82:  scaleAlg->setChild(true);

JumpFitModel.cpp:94:  extractAlg->setChild(true);

JumpFitModel.cpp:118:  appendAlg->setChild(true);

JumpFit.ui:22:     <property name="childrenCollapsible">

MSDFit.ui:22:     <property name="childrenCollapsible">
```

### PR DESCRIPTION
**Pull request by robertapplin**
**Description of work.**
The indirect inelastic GUIs can be unstable. To help minimize the problems this causes we want the GUI's to work with project recovery.
To do this the history of the output workspaces need to be fixed by:
- Making sure outputs are stored in the ADS (before being deleted at an appropriate time)
- Setting the correct algorithms to be children

**To test:**
One possible test:
1. Interfaces Menu->Indirect->Data Analysis
2. Go on I(Q, t) tab
3. Load Data
4. Click 'Run' and wait
5. Wait for Project Recovery to save and then crash mantid
6. History should have been recovered!

Fixes  #23144

Does this update require release notes?
- [ ] Yes
- [x] No

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by keeeto** [comment]
An example of a typical QENS workflow that currently fails is:

* Open Indirect > Data Analysis
* Load the file below in the MDS tab
[irs26176_graphite002_elwin_eq.nxs.zip](https://github.com/mantidproject/mantid/files/2277575/irs26176_graphite002_elwin_eq.nxs.zip)
* Set `Fit type` to Gaussian
* `Run` the fit
* After project recovery has saved files, use `Segfault` to kill Mantid
* Restart and attempt a recovery, it will fail

**Comment by keeeto** [comment]
The workflow described above still fails for the same reasons as previously. i.e. the following lines:
```
QENSFitSequential(Input='irs26176_graphite002_elwin_eq,i0;', OutputWorkspace='irs26176_graphite002_elwin_eq_MSDFit_Gaussian_s0_Result', OutputParameterWorkspace='irs26176_graphite002_elwin_eq_MSDFit_Gaussian_s0_Parameters', OutputWorkspaceGroup='irs26176_graphite002_elwin_eq_MSDFit_Gaussian_s0_Workspaces', Function='name=MsdGauss,Height=1,MSD=0.05', StartX=0.52531275787596388, EndX=1.8250942715026275, PassWSIndexToFunction=True, MaxIterations=100, IgnoreInvalidData=True)  # 2018-08-10 10:30:42.038615 
Fit(Function='name=MsdGauss,Height=0.170185,MSD=0.146884', InputWorkspace='__QENSFitSequential_ws1', IgnoreInvalidData=True, MaxIterations=100, CreateOutput=True, Output='__QENSFitSequential_ws1_0', CalcErrors=True, OutputCompositeMembers=True, StartX=0.52531275787596388, EndX=1.8250942715026275)  # 2018-08-10 10:30:42.041708 
```

The workspace `__QENSFitSequential_ws1` is hidden from the ADS, therefore doesn't exist in recovery.

To make project recovery work with Indirect workflows we need to fix this issue.

**Review by keeeto — DISMISSED** [review]
See my comments in the conversation. While this PR may help with some parts of recovering the indirect data analysis workflow it does not mean that a typical workflow is recovered.

On the plus side, it does not seem to make the interface any more unstable!

**Comment by robertapplin** [comment]
Note: The PlotPeaksByLogValueTest failed because the Fit algorithm is now recorded as a child algorithm and hence it's outputs are no longer stored in the ADS. This meant that the test that failed couldn't find the output from the Fit part of PlotPeaksByLogValue.

**Comment by Pasarus** [comment]
I spoke to @robertapplin and we believe the issue #23265 should be fixed by this PR as well.

**Comment by robertapplin** [comment]
@Pasarus It will be fixed in this specific instance within QENSFitSequential, however there may be other instances elsewhere where the behavior described in #23265 is still happening. Those instances may needed to be fixed individually as and when found

**Comment by Pasarus** [comment]
@robertapplin That is on the algorithm maintainers opposed to the people working on project recovery due to the fact that whoever wrote the algorithms in the first place making incomplete/incorrect workspace history

**Comment by robertapplin** [comment]
@Pasarus I agree that it is an issue with algorithms rather than project recovery. It's worth keeping this issue in mind however because it could cause something to not work with project recovery in the future

**Review by keeeto — DISMISSED** [review]
A lot has improved in this PR. However, I am having trouble getting this to work with ConvFit. 

* Load Mantid and got to Interfaces > Indirect > Data Analysis
* Go to the Conv Fit tab
* Load the following files into Sample and Resolution
[Archive.zip](https://github.com/mantidproject/mantid/files/2595871/Archive.zip)
* Run a fit to the data
* Wait for project saving to complete
* Crash Mantid with Segfault
* Reload Mantid and choose yes for recovery
* I see the following error:
```
RuntimeError: ['Workspace'] argument(s) not supplied to LoadParameterFile
  at line 2 in '/Users/keithbutler/.mantid/ordered_recovery.py'
```

**Comment by keeeto** [comment]
The interface now works with project recovery! Also, this passes all the stability tests for the Indirect interfaces.

**Review by keeeto — DISMISSED** [review]
This is good to go. It will be very much appreciated by the user community.

File: Framework/CurveFitting/src/Algorithms/PlotPeakByLogValue.cpp [hunk 1]
```diff
@@ -378,37 +380,33 @@ void PlotPeakByLogValue::exec() {
...
+        m_baseName + "_NormalisedCovarianceMatrices", covarianceGroup);
+
+    WorkspaceGroup_sptr parameterGroup = boost::make_shared<WorkspaceGroup>();
+    for (auto workspace : parameterWorkspaces)
+      parameterGroup->addWorkspace(workspace);
+    AnalysisDataService::Instance().addOrReplace(m_baseName + "_Parameters",
+                                                 parameterGroup);
+
+    WorkspaceGroup_sptr fitGroup = boost::make_shared<WorkspaceGroup>();
+    for (auto workspace : fitWorkspaces)
```
Comment (soininen) [inline Framework/CurveFitting/src/Algorithms/PlotPeakByLogValue.cpp]: `auto &workspace`?

File: Framework/CurveFitting/src/Algorithms/QENSFitSequential.cpp [hunk 2]
```diff
@@ -759,7 +764,14 @@ void QENSFitSequential::extractMembers(
 }
 
 void QENSFitSequential::copyLogs(
-    WorkspaceGroup_sptr resultWorkspace,
+    WorkspaceGroup_sptr resultWorkspaces,
+    const std::vector<MatrixWorkspace_sptr> &workspaces) {
+  for (const auto resultWorkspace : *resultWorkspaces)
```
Comment (soininen) [inline Framework/CurveFitting/src/Algorithms/QENSFitSequential.cpp]: `const auto &resultWorkspace`?

File: Framework/CurveFitting/src/Algorithms/QENSFitSequential.cpp [hunk 3]
```diff
@@ -772,6 +784,12 @@ void QENSFitSequential::copyLogs(
 
 void QENSFitSequential::copyLogs(MatrixWorkspace_sptr resultWorkspace,
                                  WorkspaceGroup_sptr resultGroup) {
+  for (const auto workspace : *resultGroup)
```
Comment (soininen) [inline Framework/CurveFitting/src/Algorithms/QENSFitSequential.cpp]: `&workspace`?

File: qt/scientific_interfaces/Indirect/JumpFitModel.cpp [hunk 4]
```diff
@@ -80,89 +80,101 @@ JumpFitParameters createJumpFitParameters(MatrixWorkspace *workspace) {
   return parameters;
 }
 
-MatrixWorkspace_sptr scaleWorkspace(MatrixWorkspace_sptr workspace,
-                                    double factor) {
+void deleteTemporaryWorkspaces(std::vector<std::string> const &workspaceNames) {
+  auto deleter = AlgorithmManager::Instance().create("DeleteWorkspace");
+  deleter->setLogging(false);
+  for (auto const name : workspaceNames) {
```
Comment (soininen) [inline qt/scientific_interfaces/Indirect/JumpFitModel.cpp]: `&name`?

**Review by soininen — CHANGES_REQUESTED** [review]

**Review by soininen — APPROVED** [review]

**Pull request merged**